### PR TITLE
Never give new characters unusable food or clothing (part 1)

### LINKED ISSUE #18916: Survivor starts without shoes/boots if wool allergy is chosen
We should probably provide a non-wool alternative to Winter Boots for players who choose wool allergy on survivor characters.

### PR DESCRIPTION
**Pull request by sethsimon**
Fixes #18916.

First, I wrote a test to see what starting items are affected by allergies. Here are the results, and a tentative list of item substitutions:

|  |   Wool           |       Other     |       Professions Affected |
|------------------------|---------------------------|-----------------|----------------------------|
| **blazer**          |  jacket_leather_red    |          -           |           bionic_prepper, tailor, survivalist, bionic survivalist, bionic hitman, teacher, photojournalist
| **hat_hunting** |  hat_fur                    |         -               | bow_hunter, crossbow_hunter, shotgun_hunter, rifle_hunter, lumberjack, trapper
| **hat_newsboy** | hat_cotton             |              -             |  paperboy
| **peacoat**        | jacket_flannel         |            -            | paperboy, reenactor, reenactor2
| **sweater**       | sweatshirt                |          -             | sheltered_survivor,
| **boots_winter** | boots_fur                |             -        | unemployed, paperboy
| **cloak_wool**  | cloak_leather           |             -            |  churl
| **gloves_wool** | gloves_winter         |            -                | paperboy
| **hardtack** |                    -                |    cornbread (Wheat); meat_cooked x6 (Wheat+Meatarian)       | reenactor, reenactor2
| **gum**       |                   -                |   fried_seeds (Junk) cracklins (Junk+Meatarian) |  razorgirl

I plan to create a JSON object that contains an array of objects, each of which has:
- an array of traits (if the player has these traits...)
- another array of traits (and they don't have any of these traits...)
- an old itype (then any of their starting items that have this itype...)
- a new itype (are substituted with otherwise identical item(s) that have this itype.)

If the old itype is empty, the player gets a bonus item (e.g. inhalers, glasses)- this allows the currently hardcoded item additions to be jsonized.

### CODE REVIEW

**Comment by DangerNoodle** [comment]
Can also see this latter property being use for the cannibal trait. Any other traits that come with a bonus item?

**Comment by sethsimon** [comment]
There's also the 3 types of glasses depending on nearsighted/farsighted/both, the inhaler for asthma, and an umbrella for albinos.

I'm putting this on hold for now until #19908 is merged.

**Comment by DangerNoodle** [comment]
Very well then.

**Comment by codemime** [comment]
Needs astyling.

File: src/init.cpp [hunk 1]
```diff
@@ -418,4 +420,5 @@ void DynamicDataLoader::check_consistency()
     mission_type::check_consistency();
     item_action_generator::generator().check_consistency();
     harvest_list::check_consistency();
+    profession::check_consistency_json_item_substitution();
```
Comment (codemime) [inline src/init.cpp]: I think this should be invoked from `profession::check_definitions()`.

Comment (sethsimon) [inline src/init.cpp]: Both this and `reset`, but not `load`?

Comment (codemime) [inline src/init.cpp]: `load` as well.

Comment (sethsimon) [inline src/init.cpp]: I see, I could do something like
```
"type": "profession",
"json_item_substitution": true,
"substitutions": [....
```

Comment (codemime) [inline src/init.cpp]: That's even better. I think you can also get rid of `"json_item_substitution"` and simply check if `"substitutions"` is present.

File: src/newcharacter.cpp [hunk 2]
```diff
@@ -535,32 +535,13 @@ bool player::create(character_type type, std::string tempname)
...
-        prof_items.push_back( item( "cookbook_human", 0 ) );
-    }
-    if( has_trait( "ALBINO" ) ) {
-        prof_items.push_back( item( "teleumbrella", 0 ) );
-    }
+    std::list<item> prof_items = g->u.prof->items( g->u.male, g->u.get_mutations() );
 
     for( item &it : prof_items ) {
         if( it.is_armor() ) {
-            wear_item( it, false ); // If wearing fails, we fail silently
```
Comment (kevingranade) [inline src/newcharacter.cpp]: I'm not sure failure to wear a profession item is always a bug, so I'd prefer to keep this silent.

File: src/init.cpp (hunk 1 above)
Comment (kevingranade) [inline src/init.cpp]: I suggest taking it a step further and defining the item_substitution class inside profession.cpp, nothing else has any business interacting with it, so it might as well only exist within that module.

File: src/newcharacter.cpp (hunk 2 above)
Comment (DangerNoodle) [inline src/newcharacter.cpp]: Profession generated with multiple pairs of shoes, or three copies of a clothing item? That is the only engineered instance of this situation that comes to mind

Comment (sethsimon) [inline src/newcharacter.cpp]: I plan to make the `json_item_substitution`s cover every case of wardrobe (and food) failure, and the point of that change is to note any `json_item_substitution`s that weren't added, indicating an accidental deficiency in the test case (e.g. I forgot about hooves, and the test doesn't cover any of the cases DangerNoodle mentioned).

Unless certain cases are intentionally unaccounted for by the `json_item_substitution`s, it should be treated as an error. So the question becomes, why not cover every scenario?

Comment (kevingranade) [inline src/newcharacter.cpp]: That covers most of my concern, only leaving novel combinations of traits and clothing, both of which could be added in mods.
I'm not going to strongly object though, I wasn't clear that you intend to handle every failure mode.

Comment (sethsimon) [inline src/newcharacter.cpp]: I see, the PR's title does imply that. Also, since `json_item_substitution::load` works by appending entries, mods will be able to add additional substitutions as needed.

**Comment by sethsimon** [comment]
It turns out that the answer to "why not cover any scenario?" is because adding mutations instead of only regular traits makes the test case take 317 minutes rather than 3 minutes (not a typo!). According to profiling (interrupt, backtrace, wait a few seconds, repeat) with a sample size of 8, `Character::add_traits` accounts for 100% of the execution time. I looked at its implementation, and no wonder, it loops through every trait that exists because it's impossible to get that information with any other method.

Since this is a feature branch, it would probably be out of scope to refactor that (and it would surely conflict with #19908), so I will split this into 2 parts: part 1 covers only regular traits and part 2 will do the refactoring and take care of mutations.

Also, it turns out that #19908 doesn't conflict with this like I thought it would- the new usage `scen.profquery( prof )` always gives the same result as the old usage `( scen.profsize() == 0 && !prof.has_flag( "SCEN_ONLY" ) ) || scen.profquery( prof )`.

File: src/profession.cpp [hunk 3]
```diff
@@ -24,6 +24,34 @@ generic_factory<profession> all_profs( "profession", "ident" );
 const string_id<profession> generic_profession_id( "unemployed" );
 }
 
+/**
+ * This class is used to replace certain items with other items when the player has
+ * a certain combination of traits. There is only one instance of this class, and it
+ * is only accessible by the profession module.
```
Comment (BevapDin) [inline src/profession.cpp]: > There is only one instance of this class, and it is only accessible by the profession module.

Yes, that's what the code says. Why are you repeating the code as comment?

File: src/profession.cpp [hunk 4]
```diff
@@ -44,7 +72,11 @@ profession::profession()
 
 void profession::load_profession( JsonObject &jo, const std::string &src )
 {
-    all_profs.load( jo, src );
+    if( jo.has_array( "substitutions" ) ) {
+        item_substitutions.load( jo ); // This isn't actually a profession
```
Comment (BevapDin) [inline src/profession.cpp]: This looks like a dirty hack to me. Why not make it a proper type and load it separately like all the other types?

**Comment by sethsimon** [comment]
Sorry for the delay.

>This looks like a dirty hack to me. Why not make it a proper type and load it separately like all the other types?

That would make updating `format.conf` easier, so that's probably the way to go.

File: tests/new_character_test.cpp [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+// TODO: Due to character::add_traits, mutations (ie traits unavailable in the generic scenario)
+// are commented out and not checked.
+// When they're not commented out, the test takes 317 minutes (not a typo!).
+// When they are commented out, the test takes 3 minutes.
+
+TEST_CASE( "starting_items" ) {
+    // Every starting trait that interferes with food/clothing
+    const std::vector<std::string> mutations = {
```
Comment (Coolthulhu) [inline tests/new_character_test.cpp]: Would be better if you generated that list rather than hardcoding it.
Not needed or anything, it's just that it would catch all the edge cases you could miss in manual handling.

Comment (sethsimon) [inline tests/new_character_test.cpp]: Agreed. That wouldn't be easy right now since it's all hardcoded in various places, but once the interactions between traits and food/clothing are jsonized, that would be a good enhancement.

**Comment by sethsimon** [comment]
I'm not sure what's holding up Jenkins, but this will be ready if/when the build succeeds.

**Comment by keyspace** [comment]
@sethsimon possibly [this](https://github.com/CleverRaven/Cataclysm-DDA/pull/20211/files#diff-48d9006b3f85510c6f65082fa9f651f4R224)?..

File: src/profession.h [hunk 6]
```diff
@@ -18,10 +18,10 @@ class JsonArray;
 class JsonObject;
 class addiction;
 enum add_type : int;
-    class Skill;
-    using skill_id = string_id<Skill>;
+class Skill;
+using skill_id = string_id<Skill>;
 
-    class profession
+class profession
```
Comment (narc0tiq) [inline src/profession.h]: Uh. So that's where the astyle regression came from. I mean, sure, it's basically an `astyle` bug that it's indenting here, so it's no real fault of yours.

I'll fix that real quick, now that I know what the issue is.
